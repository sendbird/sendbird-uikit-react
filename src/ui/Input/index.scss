@import '../../styles/variables';

.sendbird-input {
  display: inline-block;
  width: 100%;
  height: 40px;
  position: relative;
  .sendbird-input__input {
    width: 100%;
    height: 100%;
    border-radius: 4px;
    font-family: var(--sendbird-font-family-default);
    font-size: 14px;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    line-height: 1.43;
    letter-spacing: normal;
    padding: 10px 16px;
    box-sizing: border-box;
    background-color: inherit;
    @include themed() {
      color: t(on-bg-1);
      caret-color: t(primary-3);
      border: solid 1px t(on-bg-3);
    }
    &:focus {
      outline: none;
      @include themed() {
        box-shadow: 0 1px 5px 0 t(on-bg-4), 0 0 1px 0 rgba(0, 0, 0, 0.16), 0 2px 1px 0 rgba(0, 0, 0, 0.08);
        color: t(on-bg-1);
        border: solid 2px t(primary-3);
      }
    }
    &:active {
      @include themed() {
        box-shadow: 0 1px 5px 0 t(on-bg-4), 0 0 1px 0 rgba(0, 0, 0, 0.16), 0 2px 1px 0 rgba(0, 0, 0, 0.08);
        color: t(on-bg-1);
        border: solid 1px t(primary-3);
      }
    }
    &:disabled {
      cursor: not-allowed;
      @include themed() {
        background-color: t(bg-1);
        color: t(on-bg-3);
      }
    }
  }
  .sendbird-input__placeholder {
    position: absolute;
    pointer-events: none;
    top: 10px;
    left: 16px;
    font-family: var(--sendbird-font-family-default);
    font-size: 14px;
    line-height: 1.43;
    @include themed() {
      color: t(on-bg-3);
    }
  }
}

.sendbird-theme--light {
  --form__input__bg-color_default: var(--sendbird-light-background-50);
  
  --form__input__bg-color_disabled: var(--sendbird-dark-ondark-02);
  --form__input__bd-color_error: var(--sendbird-light-error-300);
}

.sendbird-theme--dark {
  --form__input__bg-color_default: var(--sendbird-light-onlight-03);
  
  --form__input__bg-color_disabled: var(--sendbird-light-onlight-04);
  --form__input__bd-color_error: var(--sendbird-dark-error-200);
}

.sendbird-input_for_form {
  display: inline-block;
  width: 100%;
  height: unset;
  position: relative;
  .sendbird-input__input {
    width: 100%;
    height: fit-content;
    border-radius: 4px;
    font-family: var(--sendbird-font-family-default);
    font-size: 14px;
    font-weight: normal;
    font-stretch: normal;
    font-style: normal;
    line-height: 1.43;
    letter-spacing: normal;
    padding: 7px 12px;
    box-sizing: border-box;
    background-color: inherit;
    @include themed() {
      caret-color: t(primary-3);
      color: t(inverse-on-content-1);
      background-color: var(--form__input__bg-color_default);
      border: solid 1px t(on-bg-4);
      &:disabled {
        pointer-events: none;
        border: none;
        cursor: not-allowed;
        background-color: var(--form__input__bg-color_disabled);
        color: t(on-bg-3);
      }
      &:focus {
        border: solid 1px t(primary--3-2) !important; // This is for when focus and error are being applied at the same time.
        outline: none;
        box-shadow: 0 0 0 1px t(primary--3-2);
      }
      &:active {
        box-shadow: none;
        border: solid 1px t(primary--3-2);
      }
      &.error {
        border: solid 1px var(--form__input__bd-color_error);
      }
    }
  }
  .sendbird-input__placeholder {
    position: absolute;
    pointer-events: none;
    top: 10px;
    left: 16px;
    font-family: var(--sendbird-font-family-default);
    font-size: 14px;
    line-height: 1.43;
    @include themed() {
      color: t(on-bg-3);
    }
  }
}

.sendbird-input-label {
  margin: 4px 0;
}
