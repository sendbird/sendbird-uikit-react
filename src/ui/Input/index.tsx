import React, { ReactElement, ReactNode, RefObject, useState } from 'react';

import './index.scss';

import Label, { LabelTypography, LabelColors } from '../Label';

export interface InputLabelProps {
  children: ReactNode;
}
export const InputLabel = ({ children }: InputLabelProps): ReactElement => (
  <Label
    className="sendbird-input-label"
    type={LabelTypography.CAPTION_3}
    color={LabelColors.ONBACKGROUND_1}
  >
    {children}
  </Label>
);

export interface InputProps {
  name: string;
  required?: boolean;
  disabled?: boolean;
  value?: string;
  placeHolder?: string;
  autoFocus?: boolean;
}
const Input = React.forwardRef((props: InputProps, ref: RefObject<HTMLInputElement>) => {
  const {
    name,
    required,
    disabled,
    value,
    placeHolder,
    autoFocus = false,
  } = props;
  const [inputValue, setInputValue] = useState(value);
  return (
    <div className="sendbird-input">
      <input
        className="sendbird-input__input"
        ref={ref}
        name={name}
        required={required}
        disabled={disabled}
        value={inputValue}
        onChange={(e) => {
          setInputValue(e.target.value);
        }}
        autoFocus={autoFocus}
      />
      {(placeHolder && !inputValue) && (
        <Label
          className="sendbird-input__placeholder"
          type={LabelTypography.BODY_1}
          color={LabelColors.ONBACKGROUND_3}
        >
          {placeHolder}
        </Label>
      )}
    </div>
  );
});

export default Input;
