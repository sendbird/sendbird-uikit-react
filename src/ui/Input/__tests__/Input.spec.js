import React, { useState, useRef } from 'react';
import { render, screen, fireEvent } from '@testing-library/react';

import Input from "../index";

describe('ui/Input', () => {
  it('should do a snapshot test of the Input DOM', function () {
    const { asFragment } = render(
      <Input
        value="value"
        placeHolder="placeholder"
        name="name"
        disabled={false}
        required={false}
      />,
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it('should be able to access value using ref', function () {
    const initialValue = 'initialValue';
    const finalValue = 'finalValue';

    const App = () => {
      const inputEl = useRef(null);
      const [value, setvalue] = useState("initialState");
      const onButtonClick = () => {
        // `current` points to the mounted text input element
        setvalue(inputEl.current.value)
      };
      return (
        <>
          <Input ref={inputEl} name="input" value={initialValue} />
          <button role="button" onClick={onButtonClick}>click to set value</button>
          <span id="result">{value}</span>
        </>
      );
    }
    const { container } = render(<App />);
    expect(
      screen.getByText("initialState").id
    ).toBe('result');

    fireEvent.change(container.getElementsByClassName('sendbird-input__input')[0], { target: { value: finalValue } });
    fireEvent.click(screen.getByRole('button'));
    expect(
      screen.getByText(finalValue).id
    ).toBe('result');
  });
});
