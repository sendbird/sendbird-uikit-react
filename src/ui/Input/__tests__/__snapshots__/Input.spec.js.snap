// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ui/Input should do a snapshot test of the Input DOM 1`] = `
<DocumentFragment>
  <div
    class="sendbird-input"
  >
    <input
      class="sendbird-input__input"
      name="name"
      value="value"
    />
  </div>
</DocumentFragment>
`;
